revpi_mio.c

```diff
@@ -0,0 +1,425 @@
+
+/*
+	compress the channel, only data of changed channel will be taken
+	input parameters:
+		dst, dst: compress from src to dst
+		bitmap: compress according to
+		step: number of bytes for a channel
+	return: the count of channels has been taken.
+*/
+unsigned int revpi_chnl_compress(void *dst, void *src, unsigned long bitmap, int step)
+{
+	unsigned char *d, *s;
+	unsigned int i = 0;
+
+	s = (unsigned char *)src;
+	d = (unsigned char *)dst;
+
+	for(i = 0; i < 32 && VALUE_MASK_H(bitmap, i); i++) {
+		if (test_bit(i, &bitmap)) {
+			memcpy(d, s + i * step, step);
+			d += step;
+		}
+	}
+	return (d - (unsigned char *)dst) / step;
+}
+
+int revpi_mio_cycle(unsigned char devno)
+{
+	int ret;
+	SDevice *dev;
+	struct mio_process_image_out *proc_img_out;
+	struct mio_process_image_in *proc_img_in;
+	SMioAnalogRequestData *last;
+	SMioAnalogRequestData io_req_ex;
+	unsigned int ch_cnt = 0;
+
+	dev = RevPiDevice_getDev(devno);
+	last = &mio_dio_request_last[dev->i8uPriv];
+
+	proc_img_out = (struct mio_process_image_out *)(piDev_g.ai8uPI + dev->i16uOutputOffset);
+	proc_img_in = (struct mio_process_image_in *)(piDev_g.ai8uPI + dev->i16uInputOffset);
+
+	ret = revpi_mio_cycle_dio(dev, &proc_img_out->dio, &proc_img_in->dio);
+	if (ret) {
+		return ret;
+	}
+
+	/* for the AIO cycle */
+	io_req_ex.i8uChannels = revpi_chnl_cmp(&last->i16uOutputVoltage,
+		&proc_img_out->aio.i16uOutputVoltage, MIO_AIO_PORT_CNT, 2);
+	/* force to update from process image */
+	io_req_ex.i8uChannels |= proc_img_out->aio.i8uChannels;
+
+	if(io_req_ex.i8uChannels) {
+		memcpy(&last->i16uOutputVoltage,
+				&proc_img_out->aio.i16uOutputVoltage,
+				sizeof(unsigned short) * MIO_AIO_PORT_CNT);
+
+		ch_cnt = revpi_chnl_compress(&io_req_ex.i16uOutputVoltage,
+					&proc_img_out->aio.i16uOutputVoltage,
+					io_req_ex.i8uChannels, 2);
+
+		pr_info("copy bitmap, bitmap:0x%x, copied len:%d, data:%*ph\n",
+			io_req_ex.i8uChannels, ch_cnt, 17,&io_req_ex);
+	}
+	ret = revpi_mio_cycle_aio(dev, &io_req_ex, ch_cnt, &proc_img_in->aio);
+	if (ret) {
+		return ret;
+	}
+
+	return 0;
+}
+
+int revpi_mio_reset()
+{
+	mio_conf_cnt = 0;
+
+	return 0;
+}
+
+int revpi_mio_config(unsigned char addr, unsigned short ent_cnt, SEntryInfo *ent)
+{
+	int i, offset;
+	struct mio_config *conf;
+
+	if (mio_conf_cnt >= REVPI_MIO_MAX) {
+		pr_err("max. of MIOs(%d) reached(%d)\n", REVPI_MIO_MAX, mio_conf_cnt);
+		return -ERANGE;
+	}
+
+	conf = &mio_conf_list[mio_conf_cnt];
+
+
+	revpi_io_build_header(&conf->dio.uHeader, addr, 
+				sizeof(SMioDIOConfigData), IOP_TYP1_CMD_CFG);
+
+	revpi_io_build_header(&conf->aio_i.uHeader, addr,
+				sizeof(SMioAIOConfigData), IOP_TYP1_CMD_DATA4);
+	conf->aio_i.sData.i8uDirection = 0;
+
+	revpi_io_build_header(&conf->aio_o.uHeader, addr,
+				sizeof(SMioAIOConfigData), IOP_TYP1_CMD_DATA4);
+	conf->aio_o.sData.i8uDirection = 1;
+
+	pr_info("MIO configured(addr:%d, ent-cnt:%d, conf-no:%d, conf-base:%d, \
+dio hdr:0x%x,aio_i hdr:0x%x, aio_o hdr:0x%x)\n",
```

Comment:
the idea is to break the line 269 at column 80, but if I do the intent on 270, it would produce some space in the string. what I should do to solve this?

Reply:
hm, maybe I should not break it to 2 lines, according to :"However, never break user-visible strings such as printk messages because that breaks the ability to grep for them." from linux kernel coding style